### ISSUE TITLE
Decide and implement a solution for internal-facing APIs

### ISSUE THREAD

**Issue by sffc** [unicode-org/icu4x#4467]
We've hit multiple snags recently where we wanted to change a mostly-internal function that was `pub` and were blocked because of semver:

- `cmp_iter` functions in icu_locid
- `new_owned` and `Cart` types in icu_provider
- Probably more

Some thoughts/comments/problems:

1. Even if these were private functions, we still have some desire to keep them stable-ish because then we can break hard tilde dependencies between components/utils (#4343)
2. If the functions are `#[doc(hidden)]`, they are less discoverable _including to ICU4X developers_. We can't expect ICU4X developers to look into the source code just to see if a function they want happens to be there.
3. These functions should most likely be excluded from FFI by default
4. If we are poking holes in a _util_ crate, the holes we poke should probably be part of the public API of that crate, because it is a utility crate. However, within ICU4X component and provider crates, it is okay and expected for us to poke holes.

Here's a proposal:

Add a macro_rules that generates the following boilerplate:

```rust
#[doc = "<div class=\"stab unstable\">"]
#[doc = "🚧 This code is internal to ICU4X; it may change at any time, in breaking or non-breaking ways,"]
#[doc = "including in SemVer minor releases. It can be enabled with the \"icu4x_internal\" Cargo feature."]
#[doc = "</div>"]
#[cfg(feature = "icu4x_internal")]
```

The macro_rules can live in icu_locid for now. We can move it later.

To annotate an internal API, you can do this:

```rust
/// These are docs for my internal API
icu_locid::_internal::icu4x_internal_api!()
pub struct Foo {}
```

The `icu4x_internal` Cargo feature does not bubble up to the metacrate. It can be selectively enabled whenever it is needed. One advantage of making an `icu4x_internal` feature is that we can disable it when uploading docs to docs.rs (https://docs.rs/about/metadata), although I don't think there's a way to disable _just that one feature_; we would need to switch from `--all-features` to an explicit feature list per-crate, which is error-prone. Alternatively, I'm fine if we say we don't introduce a Cargo feature for this and simply let these APIs live in the docs.

In the 1.5 or 2.0 timeframe, I suggest that we scrub our API surface and add this annotation where appropriate. A lot of things that are currently `#[doc(hidden)]` should probably use the annotation instead, and likewise things that are currently public stable should also probably use the annotation.

If we agree on this, I may suggest we make a similar annotation for the provider module.

Needs approval or input from:

- [ ] @Manishearth 
- [ ] @zbraniecki 
- [ ] @hsivonen 
- [ ] @robertbastian

**Comment by sffc** [comment]
Another solution I'm open to considering is to have the macro generate `#[doc(hidden)]`, in addition to the internal warning, and then when generating internal docs use `--document-private-items` which includes doc(hidden) APIs.

It would make me very very happy if Rust itself had an annotation for this. Anything like "this API might not be semver safe" which generates a warning when using it cross-crate.

**Comment by Manishearth** [comment]
I think what we should do for these is that we have the macro generate `cfg_attr(not(internal_docs), doc(hidden))` and generate our internal docs with `RUSTFLAGS=--cfg internal_docs`. We shouldn't use a cargo feature, we should use a cfg (since cargo features get enabled with `--all-features`)

Generating that doc header seems nice, if we use a macro we can tweak the precise solution over time.

**Comment by sffc** [comment]
@Manishearth's proposal SGTM.

Either way I would like a macro so that it's easy to find and tweak over time.

**Comment by Manishearth** [comment]
Concrete proposal:

 - We have a macro
 - It marks things as doc(hidden) except in `cfg(icu4x_internal_docs)` mode
 - ICU4X docs-gen has that CFG set
 - (optional) there's a nice "unstable feature" header


LGTM to me and Shane.

**Comment by sffc** [comment]
Specifically, the cfg should be set on the job that creates the docs uploaded to https://unicode-org.github.io/icu4x/rustdoc/, and the "unstable feature" box should be included so that people referencing those docs don't think the function is public stable.

**Comment by robertbastian** [comment]
Can we use symlinks? I'd rather not copy around, because then we'd have another CI job to enforce consistency, and another make job to "generate" these.

**Comment by sffc** [comment]
Symlinks are [annoying](https://stackoverflow.com/questions/5917249/git-symbolic-links-in-windows) for Windows users, but it appears there are [solutions](https://stackoverflow.com/questions/5917249/git-symbolic-links-in-windows/59761201#59761201) that require some manual setup.

I tend to feel that in order of priority of constituencies, the experience for new contributors should be favored over the experience of core ICU4X developers. Avoiding symlinks is harder on the core developers but easier for new contributors since they don't need any special configurations.

**Comment by robertbastian** [comment]
We've had a symlink in the repo for months now and I haven't heard any complaints.

**Comment by robertbastian** [comment]
> I tend to feel that in order of priority of constituencies, the experience for new contributors should be favored over the experience of core ICU4X developers. 

I also disagree with this. ICU4X core contributors spend a lot of time contributing to ICU4X. We just removed a bunch of generated files from the repo which again and again lead to failing CI and delays for core contributors.

**Comment by sffc** [comment]
The removal of the generated files you reference was done after establishing that doing so didn't substantially impact other constituencies.

**Comment by Manishearth** [comment]
Cross posting from https://github.com/unicode-org/icu4x/pull/5101#issuecomment-2186931612:

My opinion on this shifts if we plan to put unsafe code in `icu4x_shared`: i'd rather we used a separate crate since it seems suboptimal for every ICU4X crate to suddenly contain unsafe. This adds a ton of friction to import processes involving unsafe review; so far the fact that ICU4X is a large number of crates is manageable, but if it's a large number of crates _that includes a bunch of unsafe code_ it's worse.

**Comment by sffc** [comment]
A use case for an actual crate rather than copied-around code arose in https://github.com/unicode-org/icu4x/pull/7621.

We could just publish the crate as `icu_internal_util` (name to be bikeshed?), and we either (1) keep the version in sync with icu and exclusively use tilde dependencies, or (2) keep the version perpetually 0.x and increase it out-of-sync, more like we've been doing with icu_experimental and icu_pattern.
